trace_decoder/src/interface.rs

```diff
@@ -0,0 +1,185 @@
+pub enum SeparateTriePreImage {
+    /// Storage or state trie format that can be processed as is, as it
+    /// corresponds to the internal format.
+    Direct(HashedPartialTrie),
+}
+
+/// A trie pre-image where both state & storage are combined into one payload.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub struct CombinedPreImages {
+    /// Compact combined state and storage tries.
+    #[serde(with = "crate::hex")]
+    pub compact: Vec<u8>,
+}
+
+/// A trie pre-image where state and storage are separate.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum SeparateStorageTriesPreImage {
+    /// Each storage trie is sent over in a hashmap with the hashed account
+    /// address as a key.
+    MultipleTries(HashMap<H256, SeparateTriePreImage>),
+}
+
+/// Info specific to txns in the block.
+#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+pub struct TxnInfo {
+    /// Trace data for the txn. This is used by the protocol to:
+    /// - Mutate it's own trie state between txns to arrive at the correct trie
+    ///   state for the start of each txn.
+    /// - Create minimal partial tries needed for proof gen based on what state
+    ///   the txn accesses. (eg. What trie nodes are accessed).
+    pub traces: BTreeMap<Address, TxnTrace>,
+
+    /// Data that is specific to the txn as a whole.
+    pub meta: TxnMeta,
+}
+
+/// Structure holding metadata for one transaction.
+#[derive(Clone, Debug, Deserialize, Serialize, Default)]
+pub struct TxnMeta {
+    /// Txn byte code. This is also the raw RLP bytestring inserted into the txn
+    /// trie by this txn. Note that the key is not included and this is only
+    /// the rlped value of the node!
+    #[serde(with = "crate::hex")]
+    pub byte_code: Vec<u8>,
+
+    /// Rlped bytes of the new receipt value inserted into the receipt trie by
+    /// this txn. Note that the key is not included and this is only the rlped
+    /// value of the node!
+    #[serde(with = "crate::hex")]
+    pub new_receipt_trie_node_byte: Vec<u8>,
+
+    /// Gas used by this txn (Note: not cumulative gas used).
+    pub gas_used: u64,
+}
+
+/// A "trace" specific to an account for a txn.
+///
+/// Specifically, since we can not execute the txn before proof generation, we
+/// rely on a separate EVM to run the txn and supply this data for us.
+#[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
+pub struct TxnTrace {
+    /// If the balance changed, then the new balance will appear here. Will be
+    /// `None` if no change.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub balance: Option<U256>,
+
+    /// If the nonce changed, then the new nonce will appear here. Will be
+    /// `None` if no change.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub nonce: Option<U256>,
+
+    /// <code>[hash](keccak_hash)([Address])</code> of storages read by the
+    /// transaction.
```

Comment:
minor, but as this is part of the process of switching to support type 1 + type 2, this comment is invalid for type 2 SMT (account state keys are hashed with Poseidon, not keccak)

also applies below